### ISSUE TITLE
Move TDD module, replace with another module

### ISSUE THREAD

**Issue by vherus** [boolean-uk/software-developer#268]

**Comment by vherus** [comment]
@Auenc @Luca-Terrazzan @julesnuggy

Please have a think about potential replacements for the TDD unit and brain dump them here. There are no bad ideas at this stage, let's just brainstorm anything that might have potential and go from there.

**Comment by julesnuggy** [comment]
- **Classes (advanced)** - we only cover classes very briefly at the beginning of the BE course (first 2 weeks) and don't really delve into how to use them in more complex applications. 
  - I feel like we should cover polymorphism / extending classes too.
  - Having an exercise/challenge around using multiple classes in one application would be good. 
- **Git** - teaching them about best practices and advanced operations, such as:
  - commit strategies
  - branching
  - rebasing
  - merging

**Comment by vherus** [comment]
- How to learn new languages

**Comment by vherus** [comment]
- Authorization / roles for the auth unit

**Comment by vherus** [comment]
- Take-home interview task practice

**Comment by Auenc** [comment]
I would say, in order of priority:
- Error handling
- Async
- Design patterns

Error handling is something we really don't show a lot of. 

Async would be good to see a bit more of, but I think they get a lot of practice on the course of "the golden path". Would be good to see more error handling with async, but they would need to see more error handling in general to be able to do that comfortably.

Design patterns would be a nice to have, but I think the above two are more important to know, as any strategy they use to develop will include error handling/async

**Comment by Luca-Terrazzan** [comment]
> I would say, in order of priority:
> 
> * Error handling
> * Async
> * Design patterns
> 
> Error handling is something we really don't show a lot of.
> 
> Async would be good to see a bit more of, but I think they get a lot of practice on the course of "the golden path". Would be good to see more error handling with async, but they would need to see more error handling in general to be able to do that comfortably.
> 
> Design patterns would be a nice to have, but I think the above two are more important to know, as any strategy they use to develop will include error handling/async

I agree error handling would be really useful for them to know, as it's something they can implement tidily in their projects

Exceptions & best practices -> custom exceptions -> handling exceptions in a web app -> blankslates (404, 500 pages) -> Loggers (how to write application log for later consumptions)

Sounds like something they might be able to sell in a job interview as well

**Comment by Luca-Terrazzan** [comment]
Debugger could be something relatively easy that can help them big time during the course

**Comment by Luca-Terrazzan** [comment]
Moving to development. Will start fleshing something out on Notion and create relevant discussions there as needed

**Comment by Luca-Terrazzan** [comment]
Discussed with @vherus, we ended up with some slight updates to the existing module, which was renamed and fleshed out to not explicitly enforce TDD.

Link to Notion: https://www.notion.so/boolean/JS-software-development-basics-c6ed3d88b02149f7a0c8b25b6f224dae?pvs=4
